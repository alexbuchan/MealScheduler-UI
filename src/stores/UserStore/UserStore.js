import Store from '../Store/Store';
import Constants from '../../constants/userConstants';

let userState = {
  auth: false,
  closeFlashMessage: true
};

class UserStore extends Store {
  constructor() {
    super();
  }

  _registerToActions(action) {
    switch(action.actionType) {
      case Constants.REGISTER_USER:
        this.registerUser(action.data);
        break;

      case Constants.LOGIN_USER:
        this.loginUser(action.data);
        break;
      
      case Constants.LOGOUT_USER:
        this.logoutUser(action.data);
        break;

      case Constants.RETRIEVE_USER_DATA_ON_REFRESH:
        this.populateUserDataOnRefresh(action.data);
        break;
    }
  }

  registerUser = data => {
    this.populateUserData(data);
    this.emitChange();
  }

  loginUser = data => {
    this.populateUserData(data);
    this.emitChange();
  }

  populateUserDataOnRefresh = data => {
    this.populateUserData(data);
    this.emitChange();
  }

  logoutUser = (data) => {
    userState.user = null;
    userState.auth = false;
    userState.message = null;
    this.emitChange();
  }

  populateUserData = (data) => {
    userState.auth = data.auth;
    userState.user = data.user;
    userState.message = data.message;
  }

  getUserState() {
    return userState;
  }
}

export default new UserStore();
