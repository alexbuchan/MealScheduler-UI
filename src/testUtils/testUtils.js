import React from 'react';
import { shallow } from 'enzyme';

export const componentSetup = (Component, props = {}, options = {}) => {
  const instance = shallow(<Component { ...props } />, options);
  return instance;
}

export const findByTestAttribute = (instance, attribute) => {
  const wrapper = instance.find(`[data-test="${attribute}"]`);
  return wrapper;
}