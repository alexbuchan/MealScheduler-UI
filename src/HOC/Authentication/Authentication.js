import React from 'react';
import { Redirect } from 'react-router-dom';
import UserStore from '../../stores/UserStore/UserStore';

const withAuthentication = (WrappedComponent) => {
  return class Authentication extends React.Component {
    constructor() {
      super();
  
      this.state = {
        auth: UserStore.getUserState().auth,
      }
    }
  
    _onChange = () => {
      this.setState({
        auth: UserStore.getUserState().auth,
      });
    }
  
    componentDidMount() {
      UserStore.addChangeListener(this._onChange);
    }

    componentWillUnmount() {
      UserStore.removeChangeListener(this._onChange);
    }
  
    render() {
      if (!this.state.auth) return <Redirect to='/login' />;
      return <WrappedComponent id={ this.props.match.params.id }/>;
    }
  }
}

export default withAuthentication;