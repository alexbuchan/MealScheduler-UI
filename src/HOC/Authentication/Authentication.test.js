import Authentication from './Authentication';
import UserStore from '../../stores/UserStore/UserStore';
import { componentSetup } from '../../testUtils/testUtils';

describe('Authentication', () => {
  describe('#render', () => {
    describe('when user is NOT authenticated', () => {
      let withAuthentication;
      let instance;
    
      beforeEach(() => {
        const mockComponent = jest.fn();
        withAuthentication = Authentication(mockComponent);
        instance = componentSetup(withAuthentication);
        instance.setState({ auth: false });
      });
      
      it('should redirect to /login', () => {
        expect(instance).toMatchSnapshot();
      });
    });

    describe('when user is authenticated', () => {
      let withAuthentication;
      let instance;
    
      beforeEach(() => {
        const mockComponent = jest.fn();
        withAuthentication = Authentication(mockComponent);
        instance = componentSetup(withAuthentication);
      });

      it('should render the wrapped component', () => {
        const user = { username: 'username', email: 'user@email.com' };
        UserStore.loginUser({ user: user, message: 'User logged in.', auth: true });
        expect(instance).toMatchSnapshot();
      });
    });
  });

  describe('#componentWillUnmount', () => {
    let withAuthentication;
    let instance;
  
    beforeEach(() => {
      const mockComponent = jest.fn();
      withAuthentication = Authentication(mockComponent);
      instance = componentSetup(withAuthentication);
    });
    
    it('Should remove the Store event listener', () => {
      const removeChangeListenerSpy = jest.spyOn(UserStore, 'removeChangeListener');
      instance.unmount();
      expect(removeChangeListenerSpy).toHaveBeenCalled();
    });
  });
});