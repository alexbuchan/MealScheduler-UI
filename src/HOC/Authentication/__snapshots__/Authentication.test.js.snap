// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication #render when user is NOT authenticated should redirect to /login 1`] = `
<Redirect
  to="/login"
/>
`;

exports[`Authentication #render when user is authenticated should render the wrapped component 1`] = `<mockConstructor />`;
